File: examples/python/cart/handlers/create_cart.py
"""CreateCart command handler."""

from datetime import datetime, timezone

from google.protobuf.any_pb2 import Any
from google.protobuf.timestamp_pb2 import Timestamp

from angzarr import angzarr_pb2 as angzarr
from proto import domains_pb2 as domains
from state import CartState

from .errors import CommandRejectedError


def handle_create_cart(command_book, command_any, state: CartState, seq: int, log) -> angzarr.EventBook:
    if state.exists():
        raise CommandRejectedError("Cart already exists")

    cmd = domains.CreateCart()
    command_any.Unpack(cmd)

    if not cmd.customer_id:
        raise CommandRejectedError("Customer ID is required")

    log.info("creating_cart", customer_id=cmd.customer_id)

    event = domains.CartCreated(
        customer_id=cmd.customer_id,
        created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())),
    )

    event_any = Any()
    event_any.Pack(event, type_url_prefix="type.examples/")

    return angzarr.EventBook(
        cover=command_book.cover,
        pages=[angzarr.EventPage(num=seq, event=event_any, created_at=Timestamp(seconds=int(datetime.now(timezone.utc).timestamp())))],
    )

File: examples/rust/cart/src/handlers/create_cart.rs
//! Create cart command handler.

use angzarr::proto::{CommandBook, EventBook};
use common::proto::{CartCreated, CartState, CreateCart};
use common::{decode_command, now, require_not_exists, Result};

use crate::errmsg;
use crate::state::build_event_response;

/// Handle the CreateCart command.
///
/// Creates a new cart for a customer. Fails if cart already exists.
pub fn handle_create_cart(
    command_book: &CommandBook,
    command_data: &[u8],
    state: &CartState,
    next_seq: u32,
) -> Result<EventBook> {
    require_not_exists(&state.customer_id, errmsg::CART_EXISTS)?;

    let cmd: CreateCart = decode_command(command_data)?;

    let event = CartCreated {
        customer_id: cmd.customer_id.clone(),
        created_at: Some(now()),
    };

    Ok(build_event_response(
        state,
        command_book.cover.clone(),
        next_seq,
        "type.examples/examples.CartCreated",
        event,
    ))
}
